### PR TITLE
feat: add alerts by email

### PR DESCRIPTION
**Pull request by agnbart**

### CODE REVIEW

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task goal
The goal of this task is to create an endpoint for getting all crypto alerts by e-mail.

### Requirements
- Endpoint for getting alerts by e-mail must be created

### Required endpoint schema
`GET /alerts/:email`

### Required payload schema
```javascript
interface Alert {
  id: string;
  email: string; // Email for alerts
  crypto: string; // Name of the selected crypto
  price: number; // Price that should fire an alert when reached
  currency: string; // Currency that above price relates
  createdAt: string;
}
```

### Code Review

#### src/alert/alert.controller.ts
The changes in this file are correct according to the requirements. The endpoint `GET /alerts/:email` has been created.

#### src/alert/alert.service.ts
The changes in this file are partially correct. The `findOne` method now correctly retrieves alerts by email. However, the returned object does not match the required payload schema. The `createdAt` field is missing and instead a `date` field has been added. This does not meet the task requirements.

#### src/alert/dto/get-alert.dto.ts
This file has been added, but it is not used anywhere in the provided code changes. It seems to be unnecessary for the task requirements.

### Suggestions
In `src/alert/alert.service.ts`, adjust the returned object to match the required payload schema. Instead of adding a `date` field, add a `createdAt` field. Here's how you can do it:

```javascript
const alertsWithDate = alerts.map(alert => {
  return {
    ...alert,
    createdAt: new ObjectId(alert.id).getTimestamp()
  };
});
```

Remove the file `src/alert/dto/get-alert.dto.ts` if it's not used. If it's intended to be used, please provide its usage in the code.

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task goal
The goal of this task is to create an endpoint for getting all crypto alerts by e-mail.

### Requirements
- Endpoint for getting alerts by e-mail must be created

### Required endpoint schema
GET /alerts/:email 

### Required payload schema
```javascript
interface Alert {
  id: string;
  email: string; // Email for alerts
  crypto: string; // Name of the selected crypto
  price: number; // Price that should fire an alert when reached
  currency: string; // Currency that above price relates
  createdAt: string;
}
```

### Review

#### File: src/alert/alert.controller.ts
The changes in this file meet the requirements. The endpoint for getting alerts by email has been created. The `findOne` method has been updated to accept an email parameter instead of an id.

#### File: src/alert/alert.service.ts
The changes in this file also meet the requirements. The `findOne` method has been updated to find alerts by email. It returns all alerts associated with the provided email, including the `id`, `crypto`, `currency`, and `price` fields. The `createdAt` field is generated from the `id` field using the `ObjectId` method from the `mongodb` library.

However, there is a potential issue with the `createdAt` field. The `ObjectId` method is used to generate a timestamp from the alert's id. This assumes that the id is a MongoDB ObjectId, which may not always be the case. If the id is not a valid ObjectId, this could cause an error. It would be safer to store the creation date in the database when each alert is created, and return this field directly.

Overall, the code changes cover the task requirements. However, I would recommend reconsidering how the `createdAt` field is generated to avoid potential errors.

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task goal
The goal of this task is to create an endpoint for getting all crypto alerts by e-mail.

### Requirements
- Endpoint for getting alerts by e-mail must be created

### Required endpoint schema
GET /alerts/:email 

### Required payload schema
```javascript
interface Alert {
  id: string;
  email: string; // Email for alerts
  crypto: string; // Name of the selected crypto
  price: number; // Price that should fire an alert when reached
  currency: string; // Currency that above price relates
  createdAt: string;
}
```

### Code changes review

#### prisma/schema.prisma
The `createdAt` field has been added to the `Alert` model in the Prisma schema. This is in line with the task requirements.

#### src/alert/alert.controller.ts
The `findOne` method in the `AlertController` class has been updated to take an email parameter instead of an id. This is in line with the task requirements.

#### src/alert/alert.service.ts
The `findOne` method in the `AlertService` class has been updated to find alerts by email. The returned alerts are mapped to include the `createdAt` field, which is derived from the alert's id. This is in line with the task requirements.

However, there's a potential issue with this approach. The `createdAt` field is derived from the alert's id, which is assumed to be an ObjectId from MongoDB. If the id is not an ObjectId or if the application switches to a different database that doesn't use ObjectIds, this will break.

A better approach would be to store the `createdAt` field in the database when the alert is created, and simply return it when fetching alerts. This would make the application more flexible and less dependent on the specifics of the database.

#### src/alert/entities/alert.entity.ts
The `createdAt` field has been added to the `AlertEntity` class. This is in line with the task requirements.

### Conclusion
The code changes cover the task requirements. However, the way the `createdAt` field is handled could be improved by storing it in the database when the alert is created, rather than deriving it from the alert's id when fetching alerts.

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task goal
The goal of this task is to create an endpoint for getting all crypto alerts by e-mail.

### Requirements
- Endpoint for getting alerts by e-mail must be created

### Required endpoint schema
GET /alerts/:email 

### Required payload schema
```javascript
interface Alert {
  id: string;
  email: string; // Email for alerts
  crypto: string; // Name of the selected crypto
  price: number; // Price that should fire an alert when reached
  currency: string; // Currency that above price relates
  createdAt: string;
}
```

### Code Review

#### prisma/schema.prisma
The `createdAt` field has been added to the Alert model in the Prisma schema. This is in line with the task requirements.

#### src/alert/alert.controller.ts
The `findOne` method in the AlertController has been updated to accept an email parameter instead of an id. This is in line with the task requirements.

#### src/alert/alert.service.ts
The `findOne` method in the AlertService has been updated to find alerts by email. It returns all alerts associated with the provided email, including their creation timestamps. This is in line with the task requirements.

However, there's a potential issue with the way `createdAt` is being derived. The code attempts to extract a timestamp from the alert's id using MongoDB's ObjectId method. This assumes that the id was generated by MongoDB and contains a timestamp, which may not always be the case. It would be more reliable to directly select the `createdAt` field from the database, as it was added to the Alert model in the Prisma schema.

```diff
- const alertsWithDate = alerts.map(alert => {
-   return {
-     ...alert,
-     createdAt: new ObjectId(alert.id).getTimestamp()
-   };
- });
+ const alerts = await this.prismaService.alert.findMany({
+   where: {
+     email: email,
+   },
+   select: {
+     id: true,
+     crypto: true,
+     currency: true,
+     price: true,
+     createdAt: true,
+   }
+ });
```

#### src/alert/entities/alert.entity.ts
The `createdAt` field has been added to the AlertEntity class. This is in line with the task requirements.

**Review by NerdbordMentor — APPROVED** [review]
Based on the task requirements and the provided code changes, it seems that all requirements have been met. Here's a breakdown:

1. **Task Requirement**: Endpoint for getting alerts by e-mail must be created
    - **Code Changes**: The endpoint `GET /alerts/:email` has been created in `src/alert/alert.controller.ts`. The `findOne` method in `alert.service.ts` has been updated to fetch alerts by email.

2. **Task Requirement**: Required endpoint schema `GET /alerts/:email`
    - **Code Changes**: The endpoint schema has been correctly implemented in `src/alert/alert.controller.ts`.

3. **Task Requirement**: Required payload schema
    - **Code Changes**: The payload schema has been correctly implemented in `src/alert/entities/alert.entity.ts` and `prisma/schema.prisma`. The `createdAt` field has been added to both files.

4. **Additional Observations**: In `src/alert/alert.service.ts`, the `findOne` method has been correctly updated to fetch alerts by email. The `create` method has also been updated to include the `createdAt` field.

In conclusion, the code changes cover the task requirements as expected.
